### PR TITLE
Make `scard::Client.ioctl` return an `Encode`

### PR DESCRIPTION
**Pull request by ibeckermayer**
With the newfound confidence granted by the test suite, I changed `scard::Client.ioctl`'s signature to

```rust
fn ioctl(&mut self, code: IoctlCode, input: &mut Payload,) -> RdpResult<Option<Box<dyn Encode>>>
```

and then refactored everything to fit, including removing the inscrutable raw `Vec<u8>`'s from the tests. 

Also moves the `NTSTATUS` enum into `scard`.

### CODE REVIEW

**Comment by ibeckermayer** [comment]
Converting this to a draft because I've come across some ambiguity regarding the performance hit of switching to dynamic dispatch. 

For example, one redditor finds it to be [negligible](https://www.reddit.com/r/rust/comments/dabofa/comment/faxc5ic/?utm_source=share&utm_medium=web2x&context=3) (in our case we're doing single thousands of operations, so a 1ns difference per operation works out to single digit microseconds). But then another finds a [22x slow down](https://www.reddit.com/r/rust/comments/74llky/trait_objects_22x_slower_than_static_dispatch/). Granted the point of the many commenters who dissect why the latter benchmark test was likely exaggerated, but I'm still unclear as to how well compiler-optimized our previous iteration was.

**Comment by ibeckermayer** [comment]
> Converting this to a draft because I've come across some ambiguity regarding the performance hit of switching to dynamic dispatch.
> 
> For example, one redditor finds it to be [negligible](https://www.reddit.com/r/rust/comments/dabofa/comment/faxc5ic/?utm_source=share&utm_medium=web2x&context=3) (in our case we're doing single thousands of operations, so a 1ns difference per operation works out to single digit microseconds). But then another finds a [22x slow down](https://www.reddit.com/r/rust/comments/74llky/trait_objects_22x_slower_than_static_dispatch/). Granted the point of the many commenters who dissect why the latter benchmark test was likely exaggerated, but I'm still unclear as to how well compiler-optimized our previous iteration was.

I took some measurements to compare the old code to the new, and smartcard login took an average 594ms with the old code, and 603ms with this code (on 10 runs, data logged [here](https://docs.google.com/spreadsheets/d/1O9hUP7QNVaHG5J9zEurbNCWdRl_BYR5M4bVWkgalNw8/edit#gid=0)). Happy to consider that negligible.

**Review by mdwn — APPROVED** [review]
I'm no Rust expert, but this LGTM -- just to confirm, this is largely about readability, correct?

**Comment by ibeckermayer** [comment]
> I'm no Rust expert, but this LGTM -- just to confirm, this is largely about readability, correct?

Correct

File: lib/srv/desktop/rdp/rdpclient/src/rdpdr/scard.rs [hunk 1]
```diff
@@ -316,7 +324,15 @@ impl Client {
...
         debug!("sending {:?}", resp);
-        Ok(Some(resp.encode()?))
+        Ok(Some(Box::new(resp)))
+    }
+
+    fn handle_unimplemented_ioctl(&self, code: IoctlCode) -> RdpResult<Option<Box<dyn Encode>>> {
+        warn!("unimplemented IOCTL: {:?}", code);
+        let resp = Long_Return::new(ReturnCode::SCARD_F_INTERNAL_ERROR);
+        debug!("sending {:?}", resp);
+        // Ok(Some(resp.encode()?));
```
Comment (timothyb89) [inline lib/srv/desktop/rdp/rdpclient/src/rdpdr/scard.rs:334]: nit: can probably remove this

**Review by timothyb89 — APPROVED** [review]
Looks reasonable to me, just one trivial nit.

**Pull request merged**